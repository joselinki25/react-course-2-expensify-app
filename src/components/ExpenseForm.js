import React from 'react';
import { TextField } from '@mui/material';
import { DateTime } from 'luxon';
import { DatePicker } from '@mui/lab';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? props.expense.createdAt : DateTime.now(),
      error: '',
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount.'
      }));
    } else {
      this.setState(() => ({
        error: ''
      }));
      this.props.onSubmit({
        description: this.state.description,
        amount: Number(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt,
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        { this.state.error && <p>{this.state.error}</p> }
        <form onSubmit={this.onSubmit}>
          <input 
            type="text" 
            placeholder="Description" 
            value={this.state.description}
            onChange={this.onDescriptionChange}
            autoFocus 
          />
          <input 
            type="text" 
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <DatePicker 
            label='Created At'
            inputFormat='dd/MM/yyyy'
            value={this.state.createdAt}
            onChange={this.onDateChange}
            renderInput={(props) => <TextField {...props} />}
            showDaysOutsideCurrentMonth={false}
            inputProps={{ readOnly: true }}
          />
          <textarea 
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}