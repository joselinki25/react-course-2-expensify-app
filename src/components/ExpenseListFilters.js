import React from 'react';
import { connect } from 'react-redux';
import Button from '@mui/material/Button';
import { DateRangePicker } from '@mui/lab';
import { TextField, Box } from '@mui/material';
import { 
  setTextFilter, 
  sortByAmount, 
  sortByDate,
  setStartDate,
  setEndDate 
} from '../actions/filters';

export class  ExpenseListFilters extends React.Component{
  constructor(props) {
    super(props);
    this.state = {
      rangePickerValue: [props.filters.startDate, props.filters.endDate]
    };
  }

  onRangePickerValueChange = (newValue) => {
    this.setState(() => ({ rangePickerValue: newValue }));

    const [startDate, endDate] = newValue;

    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onTextChange = (e) => this.props.setTextFilter(e.target.value);

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div>
      <input 
        type="text" 
        value={this.props.filters.text} 
        onChange={this.onTextChange}
      />
      <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
      <DateRangePicker 
        startText="Start date"
        endText="End date"
        inputFormat="dd/MM/yyyy"
        value={this.state.rangePickerValue}
        onChange={this.onRangePickerValueChange}
        calendars={1}
        clearable={true}
        renderInput={(startProps, endProps) => (
          <React.Fragment>
            <TextField {...startProps} />
            <Box sx={{ mx: 2 }}> to </Box>
            <TextField {...endProps} />
            <Button
              onClick={() => this.onRangePickerValueChange([null, null])}
            >
              Clear
            </Button>
          </React.Fragment>
        )}
      />
    </div>
    );
  }
};

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export const ConnectedExpenseListFilter = connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);