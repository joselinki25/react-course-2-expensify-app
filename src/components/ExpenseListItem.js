import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({description, id, amount, createdAt}) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>{amount} - {createdAt && createdAt.toFormat('MMMM d, yyyy')}</p>
  </div>
)

export default ExpenseListItem;