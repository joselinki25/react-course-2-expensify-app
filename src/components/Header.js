import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <NavLink to="/" >Go home</NavLink><br/>
    <NavLink to='/create' >Create Expense</NavLink><br/>
    <NavLink to='/help' >Help</NavLink><br/>
    <br/>
  </header>
)

export default Header;