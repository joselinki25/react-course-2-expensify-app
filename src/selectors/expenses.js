// Get visible expenses

export default (
  expenses, 
  { text, sortBy, startDate, endDate }
  ) => {
    return expenses
    .filter((expense) => {
      const { createdAt } = expense;

      const startDateMatch = startDate
        ? startDate.hasSame(createdAt, 'day') || startDate < createdAt
        : true;

      const endDateMatch = endDate
        ? endDate.hasSame(createdAt, 'day') || endDate > createdAt
        : true;
        
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
  
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};