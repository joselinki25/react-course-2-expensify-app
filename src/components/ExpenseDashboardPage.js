import React from 'react';
import ExpenseList from './ExpenseList';
import { ConnectedExpenseListFilter as ExpenseListFilters } from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;