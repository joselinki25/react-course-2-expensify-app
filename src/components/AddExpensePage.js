import React from 'react'
import { connect } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';


export const AddExpensePage = (props) => {
  const navigate = useNavigate();

  const onSubmit = (expense) => {
    props.addExpense(expense);
    navigate('/');
  };

  return (
    <div>
    <h1>Add Expense</h1>
    <ExpenseForm onSubmit={onSubmit} />
    </div>
  )
}

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);