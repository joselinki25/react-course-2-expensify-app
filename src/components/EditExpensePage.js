import React from 'react'
import { useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';
import { useNavigate } from 'react-router-dom';

export const EditExpensePage = (props) => {
  let { id } = useParams();
  const navigate = useNavigate();
  const selectedExpense = props.expenses.find((expense) => expense.id === id);

  const onSubmit = (expense) => {
    props.editExpense(id, expense);
    navigate('/');
  }

  const handleRemoveExpense = () => {
    props.removeExpense(id);
    navigate('/');
  }

  return (
    <div>
      This is from my edit expense componente {id}!
      <ExpenseForm expense={selectedExpense} onSubmit={onSubmit} />
      <button onClick={handleRemoveExpense}>Remove</button>
    </div>
  )
};

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
  }
}

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (id) => dispatch(removeExpense({ id })),
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);