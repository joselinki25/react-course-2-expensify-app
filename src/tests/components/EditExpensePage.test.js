import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let expense, editExpense, removeExpense, wrapper;

beforeEach(() => {
  expense = expenses[2];
  editExpense = jest.fn();
  removeExpense = jest.fn();
  wrapper = shallow(<EditExpensePage expenses={expenses} editExpense={editExpense} removeExpense={removeExpense}/>)
})

const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate,
}));

test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expense);
  expect(editExpense).toHaveBeenLastCalledWith(expenses.id, expense);
});

test('should handle removeExpense', () => {
  wrapper.find('button').simulate('click');
  expect(removeExpense).toHaveBeenLastCalledWith(expenses.id);
});