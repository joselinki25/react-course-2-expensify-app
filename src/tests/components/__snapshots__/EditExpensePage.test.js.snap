// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  This is from my edit expense componente 
  !
  <ExpenseForm
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
